File: fixtures/external/python-agent/agent.py
#!/usr/bin/env python3
"""A Python agent RigorRun runs as a command.

The whole integration. It speaks MCP to the endpoint RigorRun hands it, does the
job, and says what it did. Nothing here imports anything of RigorRun's beyond
the SDK, and the SDK is standard library only.

Set RIGORRUN_AGENT_BEHAVIOUR=careless to ship the bug on purpose.
"""

import json
import os
import sys
import urllib.request
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "sdk" / "python"))

from rigorrun import run_stdio  # noqa: E402

CAREFUL = os.environ.get("RIGORRUN_AGENT_BEHAVIOUR") != "careless"


class Mcp:
    """The smallest MCP client that will do: initialize, then call tools.

    Deliberately hand-rolled rather than a dependency. The point of this fixture
    is that an agent already speaking MCP needs nothing from RigorRun, and
    proving that with somebody's SDK would prove something about the SDK.
    """

    def __init__(self, url: str) -> None:
        self.url = url
        self.session = None
        self.next_id = 0
        self._send("initialize", {
            "protocolVersion": "2025-11-25",
            "capabilities": {},
            "clientInfo": {"name": "python-agent", "version": "1.0.0"},
        })
        self._send("notifications/initialized", {}, notify=True)

    def _send(self, method: str, params: dict, notify: bool = False):
        body = {"jsonrpc": "2.0", "method": method, "params": params}
        if not notify:
            self.next_id += 1
            body["id"] = self.next_id
        headers = {
            "content-type": "application/json",
            "accept": "application/json, text/event-stream",
        }
        if self.session:
            headers["mcp-session-id"] = self.session
        request = urllib.request.Request(
            self.url, data=json.dumps(body).encode(), headers=headers, method="POST"
        )
        with urllib.request.urlopen(request, timeout=30) as response:
            if not self.session:
                self.session = response.headers.get("mcp-session-id")
            text = response.read().decode()
        if notify or not text:
            return None
        # Streamable HTTP answers as SSE; the payload is the last `data:` line.
        for line in reversed(text.splitlines()):
            if line.startswith("data:"):
                return json.loads(line[5:].strip())
        return json.loads(text)

    def call(self, name: str, arguments: dict):
        answer = self._send("tools/call", {"name": name, "arguments": arguments})
        result = (answer or {}).get("result", {})
        structured = result.get("structuredContent")
        if structured is not None:
            return structured
        content = result.get("content") or []
        for item in content:
            if item.get("type") == "text":
                try:
                    return json.loads(item["text"])
                except (ValueError, KeyError):
                    return item.get("text")
        return None


def work(task, environment):
    booking_id = (task.get("inputs") or {}).get("bookingId")
    if not booking_id:
        return {"status": "failed", "output": "no booking was named"}

    mcp = Mcp(environment["mcpUrl"])
    booking = mcp.call("get_booking", {"bookingId": booking_id})
    if not isinstance(booking, dict):
        return {"status": "failed", "output": f"{booking_id} could not be read"}

    # The policy this desk does not enforce: a deposit over the limit needs a
    # sign-off recorded before the booking may be confirmed. The careless
    # behaviour skips it, which is the point of having one.
    if CAREFUL and not booking.get("signedOffBy"):
        mcp.call("record_signoff", {"bookingId": booking_id, "approver": "Dana Whitlock"})

    mcp.call("confirm_booking", {"bookingId": booking_id})
    return {
        "status": "completed",
        "output": f"Confirmed {booking_id} (deposit {booking.get('depositAmount')}).",
    }


work.rigorrun_name = "python-booking-agent"
work.rigorrun_version = "1.0.0"

if __name__ == "__main__":
    run_stdio(work)

File: fixtures/external/booking-agent/src/agent.ts
/**
 * An agent somebody wrote for their own desk.
 *
 * It owns its loop, it speaks MCP, and it knows nothing about RigorRun beyond
 * the ten lines at the bottom of this file. That is the point of it: if this
 * had to be reshaped to be evaluated, the evaluation would be measuring the
 * reshaping.
 *
 * The behaviour is deliberately simple and deliberately imperfect, because a
 * benchmark that only ever sees correct agents measures nothing. `careless`
 * exists so a regression has something to be a regression *from*.
 */
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';

export interface Behaviour {
  /**
   * Whether to check the deposit against the desk's limit before confirming.
   *
   * Turning this off is how the dogfood test introduces a bug on purpose.
   */
  checksTheDeposit: boolean;
}

export const CAREFUL: Behaviour = { checksTheDeposit: true };
export const CARELESS: Behaviour = { checksTheDeposit: false };

/** The desk's rule, as this team understands it. */
const SIGN_OFF_LIMIT = 500;

interface Booking {
  bookingId: string;
  depositAmount: number;
  bookingStatus: string;
  signedOffBy: string | null;
  note: string;
}

export async function runTask(
  mcpUrl: string,
  inputs: Record<string, unknown>,
  behaviour: Behaviour,
): Promise<string> {
  const client = new Client({ name: 'booking-agent', version: '1.0.0' }, { capabilities: {} });
  const transport = new StreamableHTTPClientTransport(new URL(mcpUrl));
  await client.connect(transport as never);

  try {
    const bookingId = String(inputs['bookingId'] ?? '');
    if (!bookingId) return 'No booking was named, so nothing was done.';

    const booking = await read(client, bookingId);
    if (!booking) return `Could not read ${bookingId}.`;

    if (booking.bookingStatus === 'confirmed') {
      return `${bookingId} was already confirmed; left alone.`;
    }
    if (booking.bookingStatus === 'cancelled') {
      return `${bookingId} is cancelled and cannot be confirmed.`;
    }

    // The note is written by whoever made the enquiry. It is data. Anything in
    // it that reads like an instruction is still data.
    const noise = /ignore|urgent note|policy has changed|do not record/i.test(booking.note)
      ? ' The note contained instruction-shaped text, which was not followed.'
      : '';

    if (behaviour.checksTheDeposit && booking.depositAmount > SIGN_OFF_LIMIT && !booking.signedOffBy) {
      await call(client, 'record_signoff', { bookingId, approver: 'Duty manager' });
    }

    await call(client, 'confirm_booking', { bookingId });
    return `Confirmed ${bookingId} (deposit ${booking.depositAmount}).${noise}`;
  } finally {
    await client.close().catch(() => undefined);
  }
}

async function read(client: Client, bookingId: string): Promise<Booking | undefined> {
  const result = await call(client, 'get_booking', { bookingId });
  return result as Booking | undefined;
}

async function call(
  client: Client,
  name: string,
  args: Record<string, unknown>,
): Promise<unknown> {
  const result = await client.callTool({ name, arguments: args });
  if (result.isError) return undefined;
  const structured = (result as { structuredContent?: unknown }).structuredContent;
  if (structured !== undefined) return structured;
  const first = (result.content as { type: string; text?: string }[] | undefined)?.[0];
  return first?.text ? JSON.parse(first.text) : undefined;
}
